### PR TITLE
Migrating Maven sign and release jenkins jobs 

### LINKED ISSUE #1357: [Bug]: Maven sign and upload should be done inside a docker container
### Describe the bug

We should be using a docker container to run jenkinsjob. 

https://github.com/opensearch-project/opensearch-infra/blob/main/jenkins/jobs/OpenSearch_CI/release_ci/opensearch_maven_release/Jenkinsfile#L3-L5

Also, should this file be a part of `opensearch-build`?

### To reproduce

Docker container is not used - 

https://github.com/opensearch-project/opensearch-infra/blob/main/jenkins/jobs/OpenSearch_CI/release_ci/opensearch_maven_release/Jenkinsfile#L3-L5

### Expected behavior

_No response_

### Host / Environment

_No response_

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by zelinh**
### Description
- Move this job from Infra repo to build repo since this could use the libraries in build repo and promote our public visibility and add the unit test for this Jenkins job. 
- Add the docker CI runner image to the workflow.
- Create the function in `signArtifacts.groovy` called `generateArguments` to dynamically add arguments for the sign scripts depending on the artifacts to be signed. E.g. If it's a manifest end with `.yml`, `type` arg will be required and for other artifacts, `signatureType` arg will be required.
- Update the corresponding test cases related to the modification on `signArtifacts.groovy`
- Create a new jenkins library `downloadFromS3.groovy` and add test cases for it.
 
 
### Issues Resolved
#1357 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [inline jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:73]: Do we need to upload the Artifacts on s3? Since we are downloading them again here.

File: src/run_sign.py
Comment (abhinavGupta16) [inline src/run_sign.py:32]: Could you also change the parser to pick default in case a null or a None is passed?

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [inline jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:11]: Needs description

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [inline jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: Let's make this is a choice rather than a string input, since this would always be a fixed value of the dropdown list

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]
Needs test cases for job and lib.

Thanks for migrating the job here.

**Comment by abhinavGupta16** [comment]
ci is failing

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (zelinh) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: I would prefer we leave what it is right now. As @peternied said, we don't know what architectures we could potentially support in the future and it takes efforts to track and update choices of all jobs.

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (zelinh) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:73]: That makes sense. In the same workspace, both upload and this download are not needed.

File: src/run_sign.py
Comment (zelinh) [reply src/run_sign.py:32]: The default would not be picked at all if calling from https://github.com/opensearch-project/opensearch-build/blob/main/vars/signArtifacts.groovy, because we specified `--component=${args.component}`. In that case, even we leave the `args.component` empty or null, parser will take it as a `string` type with empty content or string `null` . 
After my change, the argument type will be set as `NoneType` so still no default would be needed.

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: Okay. I suggested a menu since it creates confusion as to what values needs to be filled in, which in this case is only `x64`. 

Out of curiosity, is it possible to have a centralized parameterized list for cases like these, params for a job?

File: src/run_sign.py
Comment (abhinavGupta16) [reply src/run_sign.py:32]: Yes, you are correct. Maybe we should change to code such that if the value being passed is 'null' or 'None', the default value gets picked automatically. That way, anyone calling the script with null or None should be able to run with defaults. Thoughts?

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [inline jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:32]: don't use `params.ARCH`, only `ARCH` would also work. using `params.ARCH` will create problems when setting variable bindings in tests

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [inline jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:14]: Unrelated thought @peternied @zelinh - do we really need this argument (`ARCH` and `PLATFORM`), given they are only used as folder paths?

Comment (zelinh) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:14]: Are you saying we could combine `ARCH` and `PLATFORM` to one parameter like called `relative path`?

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (zelinh) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:32]: I think using `param.ARCH` is more reasonable than just `ARCH`. What issue would it cause for testing?

File: src/run_sign.py
Comment (zelinh) [reply src/run_sign.py:32]: I think my way of resolving it is quite same as the circumstance you are saying. If any `null` or empty string passed in, parser will set its type to be `None`; not really necessary to have a default.

Comment (abhinavGupta16) [reply src/run_sign.py:32]: Apologies for the confusion there. I meant for other parameters as well and default value in case of `--sigType`

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:32]: `parms.ARCH` variables are not set and picked up properly when used with bindings. I faced a similar problem when writing tests for ecr.

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:14]: We could remove this parameter altogether since the value is fixed.

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: Upon further looking into this - if don't provide a dropdown list, the user can enter any value for `ARCH` and `platform`. So, one can save and create artifacts of any platform and architecture given there is no check for the values input as a string. This can be both deceiving and dangerous.

For example, I can simple enter `arm64` and the job would run fine and save artifacts as `arm64` which is incorrect.

Comment (gaiksaya) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: In that case why give an option to enter the ARCH? If we know arm64 won't work or something we do not want, `x64` is the hardcode value that can go here.

Comment (zelinh) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: Is the signer currently only working on `x64` platform? I'm not sure why other architecture is incorrect.

Comment (gaiksaya) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: We upload only x64 on maven. arm64 is out of the picture. I do not know the reason.

**Comment by abhinavGupta16** [comment]
Following up on my discussion with @peternied - The `PLATFORM` and `ARCH` parameters are required for getting the download urls for the artifacts which can we changed easily. Also, path is not fixed and any changes to the path will cause this job to fail. We should remove `PLATFORM` and `ARCH` parameters. 

We can alternatively do the following - 
- Take manifest file as input (or get one using the version number) and get the artifact path from the file directly
- Take in the artifact path as input and print the artifact path as a part of the build so it is easily available. 

I personally prefer the first one, since it truly removes manual work and automates the process, using our existing one stop shop of `manifest` files which are assured to have the most up to date values.

File: jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/release_ci/opensearch_maven_release/maven-sign-release.jenkinsfile:12]: Exactly, but this can potentially have `arm64` as input, which is incorrect

File: vars/signArtifacts.groovy
Comment (zelinh) [inline vars/signArtifacts.groovy:29]: Removed the previous modification on `run_sign.py` for circumstance of pass `None` but add the check here so that if any arguments are not specified, it wouldn't pass to the sign script.

File: vars/signArtifacts.groovy
Comment (abhinavGupta16) [inline vars/signArtifacts.groovy:38]: entry.key would not work with `--sigType`

Comment (zelinh) [reply vars/signArtifacts.groovy:38]: We didn't have a unified name for the signature type here. According to (https://github.com/opensearch-project/opensearch-build/blob/f3d8bcda4bc015786a7c167e1563fb22632ff763/src/run_sign.py#L26) we should then change the related Jenkins files arguments with arg like `sigtype: .sig` instead of `signatureType` as before.

Comment (abhinavGupta16) [reply vars/signArtifacts.groovy:38]: I think changing `opensearch-build/src/run_sign.py` would be better since `signArtifacts` is being called from a lot of places, and would require to you to test out a lot of jobs.

Comment (zelinh) [reply vars/signArtifacts.groovy:38]: That makes sense to me. Changed.

Comment (abhinavGupta16) [reply vars/signArtifacts.groovy:38]: We can also remove the function and do the following - 

```groovy

// Since the first argument should be artifactPath
String arguments = args.artifactPath

args.each(entry -> entry.key.equals('artifactPath') ? "" : arguments += " --${entry.key}=${entry.value}")

```

Comment (abhinavGupta16) [reply vars/signArtifacts.groovy:38]: adding this has only one issue, it's hard to recognize what arguments does this function take. I think it would be a good idea to add a javadoc like comment on top describing the params. WDYT?

Comment (gaiksaya) [reply vars/signArtifacts.groovy:38]: Looks like out of scope changes. @zelinh Maybe refactoring come in later PR and this PR should concentrate on maven job only?

Comment (abhinavGupta16) [reply vars/signArtifacts.groovy:38]: The maven job would require these changes to run with opensearch-build setup.

Comment (zelinh) [reply vars/signArtifacts.groovy:38]: We could add a description comment on top of this file for the instruction of this library which will tell people how the acceptable arguments.

**Comment by zelinh** [comment]
Add the unit test based on @abhinavGupta16 PR for new jenkins test framework. This is blocked by https://github.com/opensearch-project/opensearch-build/pull/1471/ until like `LibFunctionTester.groovy` and new `BuildPipelineTest.groovy` are merged.

**Comment by abhinavGupta16** [comment]
> Add the unit test based on @abhinavGupta16 PR for new jenkins test framework. This is blocked by #1471 until like `LibFunctionTester.groovy` and new `BuildPipelineTest.groovy` are merged.

Added dependency here

**Comment by abhinavGupta16** [comment]
> Add the unit test based on @abhinavGupta16 PR for new jenkins test framework. This is blocked by #1471 until like `LibFunctionTester.groovy` and new `BuildPipelineTest.groovy` are merged.

This PR is now merged. You can use the instructions from https://github.com/opensearch-project/opensearch-build/issues/1560 to write the test cases. Thanks!

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (abhinavGupta16) [inline tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:4]: Let's rename the class to `DownloadFromS3LibTester`

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (abhinavGupta16) [inline tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:39]: This variable is unused

File: vars/signArtifacts.groovy
Comment (abhinavGupta16) [inline vars/signArtifacts.groovy:11]: we could add something similar to javadocs format for all params
```groovy
/*
SignArtifacts signs the given artifacts and saves the signature in the same directory

@param Map[artifactPath] - Path to yml or artifact file
@param Map[component] <optional> - Path to yml or artifact file
*/
```

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]
Almost there! Let's add a test case for `maven-sign-release.jenkinsfile`. Also, we can get rid of `release_ci` in the directory structure to be consistent with other jobs

File: vars/signArtifacts.groovy
Comment (abhinavGupta16) [inline vars/signArtifacts.groovy:20]:
```suggestion
    
// Since the first argument should be artifactPath
String arguments = args.artifactPath

args.each(entry -> entry.key.equals('artifactPath') ? "" : arguments += " --${entry.key}=${entry.value}")

```

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (abhinavGupta16) [inline tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:22]: We can check if the value for force is a boolean here

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (abhinavGupta16) [inline tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:29]:
```suggestion
                && call.args.force.first().toString().equals(this.force)
```

File: vars/signArtifacts.groovy
Comment (gaiksaya) [reply vars/signArtifacts.groovy:41]: This change feels over complicated. Why add a loop to get all the arguments when you can get it directly form `args.variable`or maybe I am misunderstanding something?

Comment (zelinh) [reply vars/signArtifacts.groovy:41]: I was trying to iterating through the `args` without knowing what arguments would be passed in. So use the loop to iterate through the keys of the `args` map.

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (zelinh) [reply tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:4]: This file name is `TestUploadToS3LibTester` so I think these two are similar. https://github.com/opensearch-project/opensearch-build/blob/main/tests/jenkins/lib-testers/TestUploadToS3LibTester.groovy

File: vars/signArtifacts.groovy
Comment (zelinh) [reply vars/signArtifacts.groovy:20]: Ok, this seems to be more concise.

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (zelinh) [reply tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:29]: This is boolean why we check if the string is equal()? Like I define on line 9.

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (zelinh) [reply tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:39]: Will remove it.

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (abhinavGupta16) [reply tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:4]: ohh. That's my bad then. I'll rename that file as well. The other 2 are named correctly. Thanks for pointing it out!

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (abhinavGupta16) [reply tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:29]: I think everything inside `call.args.force.first()` is a string. `this.force` is boolean.

Comment (zelinh) [reply tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:29]: If that's the case, we will need something like this 
`call.args.force.first().toString().equals(this.force.toString())` to compare the actual context of string for argument `force`

File: tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy
Comment (abhinavGupta16) [reply tests/jenkins/lib-testers/TestDownloadFromS3LibTester.groovy:4]: https://github.com/opensearch-project/opensearch-build/pull/1562

File: tests/jenkins/TestMavenSignReleaseJob.groovy
Comment (abhinavGupta16) [inline tests/jenkins/TestMavenSignReleaseJob.groovy:18]: Is this the `manifestPath`?

File: tests/jenkins/TestMavenSignReleaseJob.groovy
Comment (abhinavGupta16) [inline tests/jenkins/TestMavenSignReleaseJob.groovy:20]: I think this is the `artifactPath` right? We could use this variable to set the values on line 24 and 33

File: vars/signArtifacts.groovy
Comment (abhinavGupta16) [inline vars/signArtifacts.groovy:15]: Let's remove `Choices: ['linux']` from here since this is not being tested in code.

**Review by abhinavGupta16 — COMMENTED** [review]
Suggested some minor changes in variable names for better understanding. The rest looks good to me

File: jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (zelinh) [reply jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile:32]: I could rename the `localPath` to `destPath`; for `path`, it's actually originally from Jenkins (https://www.jenkins.io/doc/pipeline/steps/pipeline-aws/#s3download-copy-file-from-s3), so I would like to keep it as what it is. WDYT?

File: tests/jenkins/TestMavenSignReleaseJob.groovy
Comment (zelinh) [reply tests/jenkins/TestMavenSignReleaseJob.groovy:18]: More precisely yes.

File: tests/jenkins/TestMavenSignReleaseJob.groovy
Comment (zelinh) [reply tests/jenkins/TestMavenSignReleaseJob.groovy:20]: Good point!

File: tests/jenkins/TestMavenSignReleaseJob.groovy
Comment (abhinavGupta16) [inline tests/jenkins/TestMavenSignReleaseJob.groovy:33]: This is `artifactPath` too

Comment (abhinavGupta16) [inline tests/jenkins/TestMavenSignReleaseJob.groovy:33]: This is `artifactPath` too.

**Review by abhinavGupta16 — APPROVED** [review]

**Review by gaiksaya — CHANGES_REQUESTED** [review]
Please run and test this workflow before merging. This jenkins job is not something that is exercised regularly so want to be sure before merging.

**Comment by zelinh** [comment]
> Please run and test this workflow before merging. This jenkins job is not something that is exercised regularly so want to be sure before merging.

I tested with upload and maven stage commented out. Should I test the uploading part?

**Comment by gaiksaya** [comment]
> > Please run and test this workflow before merging. This jenkins job is not something that is exercised regularly so want to be sure before merging.
> 
> I tested with upload and maven stage commented out. Should I test the uploading part?

Yes. It won't publish anything. It will stage the repo which we can drop later on.

File: jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (dblock) [inline jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile:30]: just echo

File: jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (dblock) [inline jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile:1]: Folders here use `-` and not `_`, so make it `opensearch-maven-release`.

File: jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile
Comment (dblock) [inline jenkins/opensearch_maven_release/maven-sign-release.jenkinsfile:14]: The description "used for artifact path" doesn't tell me anything. Do you mean "Build ID of the artifacts to stage"?.

File: vars/signArtifacts.groovy
Comment (dblock) [inline vars/signArtifacts.groovy:44]: Make a copy of `args` without `artifactsPath`, then iterate through the rest.

**Review by dblock — CHANGES_REQUESTED** [review]

File: vars/signArtifacts.groovy
Comment (zelinh) [reply vars/signArtifacts.groovy]: I updated the method that iterates through the `args` with `artifactsPath` removed.

**Comment by zelinh** [comment]
Did a test run on this job with parameters: [`BUILD_ID`: 295, `VERSION`: 1.2.3] and it ran successfully. 
Validated the artifacts are correct and signed on staging maven with @gaiksaya, and dropped after validation.

**Review by gaiksaya — APPROVED** [review]

File: src/run_sign.py
Comment (dblock) [inline src/run_sign.py:28]: To be consistent with other tools i think this should just be `platform`. I would default it to `linux` and remove the comment above.

**Review by dblock — APPROVED** [review]
LGTM. Consider renaming `distributionPlatform`. No strong feelings.

File: src/run_sign.py
Comment (zelinh) [reply src/run_sign.py]: Removed the comments and add the default value. I will leave the argument name to be `distributionPlatform` since that's what it used in some jenkinsfiles. e.g. [sign-standalone-artifacts.jenkinsfile](https://github.com/opensearch-project/opensearch-build/blob/0a20ec18c3a39770e870b2dade3c23b147962651/jenkins/sign-artifacts/sign-standalone-artifacts.jenkinsfile#L47https://github.com/opensearch-project/opensearch-build/blob/0a20ec18c3a39770e870b2dade3c23b147962651/jenkins/sign-artifacts/sign-standalone-artifacts.jenkinsfile#L47); [distribution-artifacts.jenkinsfile](https://github.com/opensearch-project/opensearch-build/blob/0a20ec18c3a39770e870b2dade3c23b147962651/jenkins/data-prepper/distribution-artifacts.jenkinsfile#L34)

File: src/run_sign.py
Comment (dblock) [inline src/run_sign.py:27]: Actually I think we should change this to match the other tools. Jenkins argument is beyond the point, but we we don't use camelCase elsewhere, e.g. https://github.com/opensearch-project/opensearch-build/blob/e45558ecca8c190f694d1fa52707a184d999c3c3/src/assemble_workflow/assemble_args.py#L19.

So I think this should be back to `--sigtype` and `--platform` or we're introducing a different pattern in Python tools.

**Review by dblock — CHANGES_REQUESTED** [review]
See below.

File: src/run_sign.py
Comment (zelinh) [reply src/run_sign.py:27]: @abhinavGupta16 Do you have any objections we moving this back to `sigtype`? 
We would also need to change the arguments in quite some places where `signArtifacts.groovy` is called.

**Comment by zelinh** [comment]
Name convention for the type of signature (e.g. `.sig`) has been changed from `signatureType` to `sigtype` across all jenkins job calling `signArtifacts` to be consistent with other tools as we didn't use camelCases for our arguments or variables.

**Review by dblock — APPROVED** [review]
This works for me.

Generally the boundary should be at interface points, so I don't think you needed to change the convention in groovy code which generally uses camelCase, only in python.

**Pull request merged**
